man/sslclient.1
.TH sslclient 1
.SH NAME
sslclient \- setup a TLS client connection
.SH SYNOPSIS
.B sslclient 
[
.I opts
]
.I host 
.I port 
.I prog 
.SH DESCRIPTION
.I opts 
is a series of getopt-style options,
.I host 
is a host name for the client to connect to, and
.I prog 
is one or more arguments specifying a program to run 
for each successful connection. 

.B sslclient 
attempts to connect to a TCP server at 
.I host
.IR port . 
The server's address is given by
.I host
and
.IR port .
.I host
may be
.IR 0 ,
which is identical to
.IR 127.0.0.1 ,
or
.IR ::1
referring to the IPv6 loopback address,
a compactified IPv6 address,
a dotted-decimal IPv4 address,
or a host name.
If a host name is given,
.B sslclient
facilitates a DNS lookup and tries each address in turn.

If the connection succeeds, 
.B sslclient 
runs 
.IR prog , 
with file descriptors 6 and 7 reading from and 
writing to a child process ssl. 
The ssl process attempts an SSL connect via the network. 
If it succeeds, it translates data between 
.I prog 
and the network, performing any necessary SSL encoding and decoding. 

Before running 
.I prog, 
.B sslclient 
sets certain environment variables. 
.SH OPTIONS
.B General Options: 
.TP
.B \-q
Quiet. Do not print error messages. 
.TP
.B \-Q
(Default.) Print error messages. 
.TP
.B \-v 
Verbose. Print error messages and status messages.
.P
.B Connection options: 
.TP
.B \-4
Use IPv4 sockets for connections and DNS queries. 
.TP
.B \-6
Force IPv6 mode for connections.
.TP
.B \-T \fIx+y
Give up on the connection attempt or SSL connection attempt after 
.I x+y 
seconds. The default value is: 2+58. When a 
.I host 
has several IP addresses,
.B sslclient 
tries to connect to the first IP address, waits 
.I x 
seconds, tries to connect to the second IP address, waits again
.I x 
seconds, etc.; then it retries each address that timed out, waiting 
.I y 
seconds per address. You may omit 
.I +y 
to skip the second try. 
.TP
.B \-i \fIlocalip
Use 
.I localip 
as the IP address for the local side of the connection; quit if 
.I localip 
is not available. Normally 
.B sslclient 
lets the operating system choose an address. 
.TP
.B \-p \fIlocalport
Use 
.I localport
as the TCP port for the local side of the connection; quit if 
.I localport 
is not available. Normally 
.B sslclient 
lets the operating system choose a port. 
.TP
.B \-d
Delay sending data for a fraction of a second whenever 
the remote host is responding slowly. 
This is currently the default, but it may not be in the future; 
if you want it, set it explicitly. 
.TP 
.B \-D
Never delay sending data; enable TCP_NODELAY. 
.TP
.B \-I \fIifname
Use
.I ifname
as the local network interface.  This is only defined for IPv6 sockets
and needed if you use link-local IPv6 addresses.
.TP
.B \-M
(Default.) No Server Name Indications are present.
.TP
.B \-m
Employ Server Name Indication (SNI) for 
the given
.I hostname
in the Client Helo.

\fINote\fR: You can use \fIDNSCACHEIP\fR to set the DNS resolver IP dynamically.

.P
.B X509 certificate handling:
.TP
.B \-3
Read a null-terminated 
.I key password 
from file descriptor 3 or alternatively,
.TP
.B \-K \fIkeypassfile
Read a null-terminated
.I key password
from 
.IR keypassfile .
.TP
.B \-a \fIcafile
Override the compiled-in CA file name. 
The CA file contains the list of CAs used to verify the server certificate. 
.TP
.B \-A \fIcadir
Override the compiled-in CA directory name. 
The CA directory contains certificates files used 
to verify the client certificate. This list augments the list from 
.I \-a \fIcafile. 
.TP
.B \-c \fIcertfile
Use the client certificate in 
.IR certfile . 
.TP
.B \-k \fIkeyfile
Use the client certificate key in 
.IR keyfile . 
.TP
.B \-V \fIverifydepth
Verify the server certificate chain to depth 
.IR verifydepth . 
The default value is 1. 
.TP
.B \-z \fIciphers 
Use the cipher list specified in 
.IR ciphers . 
.TP
.B \-x
(Default.) Verify the server certificate. 
.TP
.B \-X 
Do not verify the server certificate.
.TP
.B \-n 
(Default.) Verify that the server host name matches 
the FQDN provided in the certificate.
.TP
.B \-N
Do not verify that the server host name matches 
the FQDN provided in the certificate.
.P
.B Data-gathering options: 
.TP
.B \-h
(Default.) Look up the remote host name in DNS 
to set the environment variable $SSLREMOTEHOST. 
.TP
.B \-H
Do not look up the remote host name in DNS; 
remove the environment variable $SSLREMOTEHOST. 
.TP
.B \-l \fIlocalname
Do not look up the local host name in DNS; use 
.I localname 
for the environment variable $SSLLOCALHOST. 
A common choice for
.I localname 
is
.I 0
or
.IR :: . 
.TP
.B \-r 
Attempt to obtain $SSLREMOTEINFO from the remote host. 
.TP
.B \-R 
(Default.) Do not attempt to obtain $SSLREMOTEINFO from the remote host. 
.TP
.B \-t \fIn
Give up on the $SSLREMOTEINFO connection attempt after 
.I n 
seconds. The default value is: 26. 
.TP
.B \-w \fIn
Give up on a connection or program after waiting 
.I n 
seconds for read or write. The default value is: 3600. 
.TP
.B \-s 
Store client and server certificate information in the environment, a la mod_ssl.
.TP 
.B \-S
(Default.) Do not store client and server certificate information in the environment. 
.TP
.B \-e 
Set protocol environment a la 
.BR tcpserver . 
Set $PROTO, $TCPLOCALIP, $TCPLOCALPORT, $TCPLOCALHOST, $TCPREMOTEIP, 
$TCPREMOTEPORT, $TCPREMOTEHOST, and $TCPREMOTEINFO 
from the corresponding $SSL variables. 
.TP
.B \-E 
(Default.) Do not set any 
.B tcpserver 
environment variables.
.SH SEE ALSO
sslserver(1),
sslhandle(1),
sslconnect(1),
sslcat(1),
https@(1),
ucspi-tls(2),
tcpclient(1),
tcpserver(1),
tcp-environ(5).
.SH REFERENCE
http://httpd.apache.org/docs/2.4/mod/mod_ssl.html


doc/CHANGELOG
20011209
	Initial release.
	version: 0.50

20021207
	Convert to new build project.
	Update libraries.
	Correct load ordering for ssl programs.

20030118
	Option to set TCP environment variables.
	Option to set SSL environment variables a la mod_perl.
	Added sslperl.
	Correct closure handling in ssl_io.c.

20030128
	Correct setting of TCPREMOTEINFO.

20040127
	Improved env tracking in handlers.
	Updated tests.

20040204
	Clean up signal handling, process usage.

20040209
	Added nN options to sslclient.

20040313
	Added conf-ssl.
	Version 0.65.

20040317
	Delete conf-home.
	Permit nN opts in sslclient.
	Version 0.66.

20040320
	Allow sslclient to avoid server verification.
	Support cipher selection.
	Correct execution botch in sslperl.
	Clean up ssl_io and signal interaction.
	Version 0.67.

20040502
	Read SSL setup info before dropping privilege.
	Version 0.68.

20050417
	Don't fail on setsid if already group leader (Gabriel Russell).
	Add compatibility macros for older Perl.
	Don't leak pointer returned from eval_pv.
	Support for partial testing.
	Remove dependency on daemontools from rts.

20050515
	Partial-install bug fix in package/upgrade.

20050605
	Don't lose file descriptors in sslclient (Paul Jarc).

20050610
	Close socket for deny rule (Fred Lindberg).

20050703
	Revise ssl_io.
	Free ssl object in sslhandle.c.
	Add progtimeout option.

20050717
	Version 0.70.

20090811 
	Included ucspi-ssl-0.70_ucspitls-0.6.patch (STARTTLS support)
	originally designed and provided by Scott Gifford (FEH).

20100319
	Added Certchain support for sslserver and sslclient (FEH).

20120217
	Integration and added man-pages (FEH).
	Version 0.80.

20120921
	Synced with ucspi-tcp6-0.95.
	Version 0.82

20121003
	Fixed integration bug in ssl_very.c.
	Included patches from Peter Conrad.
	Version 0.83.

20121005
	Bug fix in sslserver. Several small
	corrections.
	Version 0.83a

20121019
	Fix for large X509 serial numbers on x86 (tx. Peter Conrad).
	SAN DNSname has precedence over CN in subject.
	Re-edited man pages and rts tests.
	Version 0.84.

20130602
	Added IPv6 support (tx. to Felix von Leitner and Brandon Turner).
	UI: Changed sslserver client cert call from '-i/-I' to '-z/-Z'
	for compatibility reasons.
	Added '-4/-6' support for client scripts.  
	Version 0.90. 

20130804
	Added output environment variables TCP6* for sslserver.
	sslperl, sslhandle, and sslprint are not IPv6 ready yet. 
	Version 0.91 test.

20130910
	Added IPv6 capabilities to sslhandle, sslprint, sslperl.
	Changed verification of X.509 certs. 
	Removed obsolete socket_4 calls in sslserver.
	Version 0.92 beta.

20140112
	Streamlined code with ucspi-tcp6-1.00.
	Supplied new certs with customized SAN.
	Make rts working (at least some how).

20140331
	Added support for personalized client certs.
	New option '-m' in sslserver, complementing '-z'.
	CCAFILE='-' disables client cert request.
	Version 0.94 beta.

20141221
	Added verbose log output for SSL connection informations.
	Version 0.95a beta. 

20140208
	Fixed wrongly nested CONNECT error code for sslclient.c
	producing wrong warning messages while connecting to
	an IPv4 address. 
	Added call of '-ldl' in ssl.lib.
	Version 0.95b beta.

20151101
	Mitigation of SSL connection hanging during
	coincident change of daylight-saving settings. 

20160228
	Fixed bug in sslserver's dnsip lookup in case of paranoid settings 
	and additonal existance of IPv6 AAAA records for incoming IPv4 connection.
	Version 0.96.

20160802
	Serveral fixes from 'troy@' included to cope with compiler errors and 
	to solve a bug in function getbitasaddress in ip4_bit.c (= ucspi-tcp6-1.02).
	Reordered conf-* variables in main dir to allow easier generation of 
	packages (i.e. RPM). Fixed script to identify different HW architecture
	and OS. This version works in 32 bit mode on Raspian Linux / RasPi 7. 

 	Added ECDH capabilites (tx to Frank Bergmann for the patches).
	Version 0.97.

20161226
	Added compatibility with LibreSSL.
	Fixed missing negative return call treatment from 'poll' (tx Frank Bergmann).
	Tentative 'emake' fix for Gentoo build.
	Version 0.98a.

20170209
	Added OpenSSL 1.1 tweaks -- works under Debian (9) 'Stretch'.
	Version 0.99.

20170308
	Included PID in sslserver + sslhandle abend logs in case of SSL failure.
	Version 0.99a.

20170617
	Convenience release: Removed references to 'gcc' and used 'cc' instead.
	Version 0.99b.

20170712
	Convenience release: Added `correct` pid display in error log.

20171028
	Fixed cosmetic bug in sslserver displaying parent and not child pid in log.
	Tx Bruce Guenter.

20171105
	Clean ups.

20180811
	Fixed missing 'return 0' in ssl_params.c for ECDH handshake (tx. J.W.).
	Version 0.99e

----

20180809
	Complete refurbish based on fehQlibs.
	Native handling of IPv4/IPv6 address for sslclient.
	Version 0.10

20180810
	Added experimental 'ecdhparam' file.
	Version 0.10.1

20180816
	Removed experimental 'ecdhparam' handling -- OpenSSL does not support it.
	fehQlibs-08 required.
	Version 0.10.2

20181010
	Finished TLS 1.3 integration (based on OpenSSL 1.1.1).
	Removed compiler flags for ECDH -- now required.
	fehQlibs-09 based.
	Version 0.10.6

20181109
	Better handling of read EAGAIN (sslserver may hang).
	Include socket_dualstack option (required for OS with IPv6_V6ONLY).
	fehQlibs-10 required.
	Version 0.10.7

20190318
	Added dualstack handling for servers applying the 
	pseudo IP address ':0' on call (common now for all servers).
	Tailored TLS error handling for EAGAIN end error codes.
	Rewrote IPv4 CIDR address evaluation for rules.
	Version 0.10.8

20190505
	Fixed broken evaluation of CIDR and IPv6 addresses; 
	adjusted with ucspi-tcp6-1.10.5.
	Improved compatibility with LibreSSL and included description.
	Version 0.10.9

20190608
	Added DSA/DSS (+ECC) signature verification additionally to RSA.
	Added compatibility with fehQlibs-12.
	Version 0.10.10

20190728
	Compatibility improvements for the forthcoming s/qmail.

src/sslclient.c
/** 
  @file  sslclient.c
  @author web, fefe, feh
  @brief IPv6 enabled sslclient
*/
#include <unistd.h>
#include <open.h>
#include <sys/types.h>
#include <sys/param.h>
#include <netdb.h>
#include <arpa/inet.h>
#include "ucspissl.h"
#include "sig.h"
#include "exit.h"
#include "getoptb.h"
#include "uint_t.h"
#include "fmt.h"
#include "scan.h"
#include "str.h"
#include "ip.h"
#include "socket_if.h"
#include "fd.h"
#include "stralloc.h"
#include "buffer.h"
#include "getln.h"
#include "logmsg.h"
#include "pathexec.h"
#include "timeoutconn.h"
#include "remoteinfo.h"
#include "dnsresolv.h"
#include "byte.h"
#include "ndelay.h"
#include "wait.h"
#include "auto_cafile.h"
#include "auto_cadir.h"
#include "auto_ciphers.h"

#define WHO "sslclient"

void nomem(void) {
  logmsg(WHO,111,FATAL,"out of memory");
}
void env(const char *s,const char *t) {
  if (!pathexec_env(s,t)) nomem();
}

void usage(void) {
  logmsg(WHO,100,USAGE,"sslclient \
[ -463hHrRdDiqQveEsSnNxX ] \
[ -i localip ] \
[ -p localport ] \
[ -T timeoutconn ] \
[ -l localname ] \
[ -t timeoutinfo ] \
[ -I interface ] \
[ -a cafile ] \
[ -A cadir ] \
[ -c certfile ] \
[ -z ciphers ] \
[ -k keyfile ] \
[ -K keypassfile ] \
[ -V verifydepth ] \
[ -w progtimeout ] \
host port program");
}

int verbosity = 1;
int flagdelay = 0;
int flagremoteinfo = 0;
int flagremotehost = 1;
int flag3 = 0;
int flagsslenv = 0;
int flagtcpenv = 0;
int flagsni = 0;
unsigned long itimeout = 26;
unsigned long ctimeout[2] = { 2, 58 };
unsigned int progtimeout = 3600;
uint32 netif = 0;

const char *loopback = "127.0.0.1";
char iplocal[16] = { 0,0,0,0, 0,0,0,0, 0,0,0,0, 0,0,0,0 };
uint16 portlocal = 0;
const char *forcelocal = 0;

char ipremote[16];
uint16 portremote;

const char *hostname;
int flagname = 1;
int flagservercert = 1;
static stralloc addresses;
static stralloc certname;
static stralloc moreaddresses;

static stralloc tmp;
static stralloc fqdn;
static char strnum[FMT_ULONG];
static char ipstr[IP6_FMT];

char seed[128];

char bspace[16];
buffer b;

SSL_CTX *ctx;
const char *certfile = 0;
const char *keyfile = 0;
const char *keypass = 0;
const char *cafile = auto_cafile;
const char *cadir = auto_cadir;
const char *ciphers = auto_ciphers;
stralloc password = {0};
int match = 0;
int verifydepth = 1;

int pi[2];
int po[2];
int pt[2];

void read_pwdfile()
{
  int fd;
  
  if (!password.len) {
    fd = open_read(keypass);
    if (fd == -1) 
      logmsg(WHO,111,ERROR,B("unable to read password from: ",keypass));

    buffer_init(&b,buffer_unixread,fd,bspace,sizeof(bspace));
    if (getln(&b,&password,&match,'\0') == -1)
      logmsg(WHO,111,ERROR,B("unable to read password from: ",keypass));
    close(fd);
    if (match) --password.len;
  }
}

void read_passwd() {
  if (!password.len) {
    buffer_init(&b,buffer_unixread,3,bspace,sizeof(bspace));
    if (getln(&b,&password,&match,'\0') == -1)
      logmsg(WHO,111,ERROR,"unable to read password from FD3");
    close(3);
    if (match) --password.len;
  }
}

int passwd_cb(char *buf,int size,int rwflag,void *userdata) {
  if (size < password.len)
    logmsg(WHO,111,ERROR,"password too long");

  byte_copy(buf,password.len,password.s);
  return password.len;
}

int main(int argc,char * const argv[]) 
{
  unsigned long u;
  int opt;
  const char *x;
  int j;
  int s;
  int r;
  int cloop;
  SSL *ssl;
  int wstat;
  int ipflag = 0;

  dns_random_init(seed);

  close(6);
  close(7);
  sig_ignore(sig_pipe);
 
  while ((opt = getoptb(argc,(char **)argv,"dDvqQhHrRimM:p:t:T:l:a:A:c:z:k:K:V:346eEsSnN0xXw:")) != opteof)
    switch(opt) {
      case '4': ipflag = 1; break;
      case '6': ipflag = 0; break;
      case 'd': flagdelay = 1; break;
      case 'D': flagdelay = 0; break;
      case 'm': flagsni = 1; break;
      case 'M': flagsni = 0; break;
      case 'v': verbosity = 2; break;
      case 'q': verbosity = 0; break;
      case 'Q': verbosity = 1; break;
      case 'l': forcelocal = optarg; break;
      case 'H': flagremotehost = 0; break;
      case 'h': flagremotehost = 1; break;
      case 'R': flagremoteinfo = 0; break;
      case 'r': flagremoteinfo = 1; break;
      case 't': scan_ulong(optarg,&itimeout); break;
      case 'T': j = scan_ulong(optarg,&ctimeout[0]);
                if (optarg[j] == '+') ++j;
                scan_ulong(optarg + j,&ctimeout[1]); break;
      case 'w': scan_uint(optarg,&progtimeout); break;
      case 'i': if (!ip6_scan(optarg,iplocal)) usage(); break;
      case 'I': netif = socket_getifidx(optarg); break;
      case 'p': scan_ulong(optarg,&u); portlocal = u; break;
      case 'a': cafile = optarg; break;
      case 'A': cadir = optarg; break;
      case 'c': certfile = optarg; break;
      case 'z': ciphers = optarg; break;
      case 'k': keyfile = optarg; break;
      case 'K': keypass = optarg; break;
      case 'V': scan_ulong(optarg,&u); verifydepth = u; break;
      case '3': flag3 = 1; break;
      case 'S': flagsslenv = 0; break;
      case 's': flagsslenv = 1; break;
      case 'E': flagtcpenv = 0; break;
      case 'e': flagtcpenv = 1; break;
      case 'N': flagname = 0; break;
      case 'n': flagname = 1; break;
      case 'x': flagservercert = 1; break;
      case 'X': flagservercert = 0; break;
      default: usage();
    }
  argv += optind;

  if (!verbosity)
    buffer_2->fd = -1;

  hostname = *argv;
  if (!hostname || str_equal((char *)hostname,"")) usage();
  if (str_equal((char *)hostname,"0")) hostname = loopback;

  x = *++argv;
  if (!x) usage();
  if (!x[scan_ulong(x,&u)])
    portremote = u;
  else {
    struct servent *se;
    se = getservbyname(x,"tcp");
    if (!se)
      logmsg(WHO,111,FATAL,B("unable to figure out port number for ",x));
    uint16_unpack_big((char*)&se->s_port,&portremote);
  }

  if (flag3) read_passwd();
  if (keypass) read_pwdfile();

  if (cafile && str_equal(cafile,"")) cafile = 0;
  if (cadir && str_equal(cadir,"")) cadir= 0;
  if (ciphers && str_equal(ciphers,"")) ciphers= 0;

  if (certfile && str_equal(certfile,"")) certfile = 0;
  if (keyfile && str_equal(keyfile,"")) keyfile = 0;
  if (keypass && str_equal(keypass,"")) keypass = 0;

  if (!*++argv) usage();

  /* IP address only */

  if (ip4_scan(hostname,ipremote)) {
    if (!stralloc_copyb(&addresses,(char *)V4mappedprefix,12)) nomem();
    if (!stralloc_catb(&addresses,ipremote,4)) nomem();
    byte_copy(iplocal,16,addresses.s);
  } else if (ip6_scan(hostname,ipremote)) {
      if (!stralloc_copyb(&addresses,ipremote,16)) nomem();
      byte_copy(iplocal,16,addresses.s);
  }

  if (addresses.len < 4) {
    if (!stralloc_copys(&tmp,hostname)) nomem();
    dns_ip_qualify(&addresses,&fqdn,&tmp);
    if (addresses.len < 16) 
      logmsg(WHO,111,ERROR,B("No IP address for: ",hostname));
  }

  if (addresses.len == 16) {
     ctimeout[0] += ctimeout[1];
     ctimeout[1] = 0;
  }

  for (cloop = 0; cloop < 2; ++cloop) {
    if (!stralloc_copys(&moreaddresses,"")) nomem();
    for (j = 0; j + 16 <= addresses.len; j += 16) {
      if (ipflag == 1 || ip6_isv4mapped(addresses.s + j)) {
        s = socket_tcp4();
        if (s == -1) logmsg(WHO,111,FATAL,"unable to create socket");
        r = socket_bind4(s,iplocal,portlocal);
      } else {
        s = socket_tcp6();
        if (s == -1) logmsg(WHO,111,FATAL,"unable to create socket");
        r = socket_bind6(s,iplocal,portlocal,netif);
      }   
      strnum[fmt_ulong(strnum,portlocal)] = 0;
      if (r == -1) logmsg(WHO,111,FATAL,B("unable to bind to socket for local port: ",strnum));
      if (timeoutconn(s,addresses.s + j,portremote,ctimeout[cloop],netif) == 0)
        goto CONNECTED;
      close(s);
      if (!cloop && ctimeout[1] && (errno == ETIMEDOUT)) {
        if (!stralloc_catb(&moreaddresses,addresses.s + j,16)) nomem();
      }   
      else {
        strnum[fmt_ulong(strnum,portremote)] = 0;
        if (ip6_isv4mapped(addresses.s + j)) 
          ipstr[ip4_fmt(ipstr,addresses.s + j + 12)] = 0;
        else
          ipstr[ip6_fmt(ipstr,addresses.s + j)] = 0;
      }   
    }   
    if (!stralloc_copy(&addresses,&moreaddresses)) nomem();
  }
  logmsg(WHO,110,DROP,B("unable to connect to: ",ipstr," port: ",strnum));

  _exit(111);

  CONNECTED:

  /* Local */

  if (socket_local(s,iplocal,&portlocal,&netif) == -1)
    logmsg(WHO,111,FATAL,"unable to get local address");

  if (ip6_isv4mapped(iplocal)) {
    env("PROTO","TCP6");
    ipstr[ip4_fmt(ipstr,iplocal + 12)] = 0;
  } else {
    env("PROTO","TCP6");
    if (flagtcpenv && netif) env("TCP6INTERFACE",socket_getifname(netif));
    ipstr[ip6_fmt(ipstr,iplocal)] = 0;
  }

  env("SSLLOCALIP",ipstr);
  if (flagtcpenv) env("TCPLOCALIP",ipstr);

  strnum[fmt_ulong(strnum,portlocal)] = 0;
  env("SSLLOCALPORT",strnum);
  if (flagtcpenv) env("TCPLOCALPORT",strnum);

  x = forcelocal;
  if (!x)
    if (dns_name(&tmp,iplocal) >= 0) {
      if (!stralloc_0(&tmp)) nomem();
      x = tmp.s;
    }
  env("SSLLOCALHOST",x);
  if (flagtcpenv) env("TCPLOCALHOST",x);

  /* Remote */

  if (socket_remote(s,ipremote,&portremote,&netif) == -1)
    logmsg(WHO,111,FATAL,"unable to get remote address");

  if (ip6_isv4mapped(ipremote)) 
    ipstr[ip4_fmt(ipstr,ipremote + 12)] = 0;
  else
    ipstr[ip6_fmt(ipstr,ipremote)] = 0;

  env("SSLREMOTEIP",ipstr);
  if (flagtcpenv) env("TCPREMOTEIP",ipstr);

  strnum[fmt_ulong(strnum,portremote)] = 0;
  env("SSLREMOTEPORT",strnum);
  if (flagtcpenv) env("TCPREMOTEPORT",strnum);

  x = 0;
  if (flagremotehost)
    if (dns_name(&tmp,ipremote) >= 0) {
      if (!stralloc_0(&tmp)) nomem();
      x = tmp.s;
    }

  env("SSLREMOTEHOST",x);
  if (flagtcpenv) env("TCPREMOTEHOST",x);

  x = 0;
  if (flagremoteinfo)
    if (remoteinfo(&tmp,ipremote,portremote,iplocal,portlocal,itimeout,netif) == 0) {
      if (!stralloc_0(&tmp)) nomem();
      x = tmp.s;
    }
  env("SSLREMOTEINFO",x);
  if (flagtcpenv) env("TCPREMOTEINFO",x);

  /* Context */

  ctx = ssl_client();
  ssl_errstr();
  if (!ctx)
    logmsg(WHO,111,FATAL,"unable to create TLS context");

  switch (ssl_certkey(ctx,certfile,keyfile,passwd_cb)) {
    case -1: logmsg(WHO,111,ERROR,"unable to load certificate");
    case -2: logmsg(WHO,111,ERROR,"unable to load key pair");
    case -3: logmsg(WHO,111,ERROR,"key does not match certificate");
    default: break;
  }
  
  if (flagservercert && !ssl_ca(ctx,cafile,cadir,verifydepth))
    logmsg(WHO,111,ERROR,"unable to load CA list");

  if (!ssl_ciphers(ctx,ciphers))
    logmsg(WHO,111,ERROR,"unable to set cipher list");

  ssl = ssl_new(ctx,s);
  if (!ssl) logmsg(WHO,111,FATAL,"unable to create TLS instance");

  if (flagsni)
    if (!SSL_set_tlsext_host_name(ssl,hostname))
      logmsg(WHO,111,FATAL,B("unable to set TLS SNI extensions for hostname: ",(char *)hostname));

  for (cloop = 0; cloop < 2; ++cloop) {
    if (!ssl_timeoutconn(ssl,ctimeout[cloop])) goto SSLCONNECTED;
    if (!cloop && ctimeout[1]) continue;
    logmsg(WHO,111,FATAL,"unable to TLS connect");
  }

  _exit(111);

  SSLCONNECTED:

  ndelay_off(s);

  if (flagservercert)
    switch(ssl_verify(ssl,hostname,&certname)) {
      case -1:
        logmsg(WHO,110,ERROR,"no server certificate");
      case -2:
        logmsg(WHO,110,ERROR,"missing credentials (CA) or unable to validate server certificate");
      case -3:
        if (!stralloc_0(&certname)) nomem();
        if (flagname) 
          logmsg(WHO,110,ERROR,B("server hostname does not match certificate: ",(char *)hostname," <=> ",certname.s));
      default: break;
    }

  if (verbosity >= 2)
    log_who(WHO,B("tls connected to: ",ipstr," port: ",strnum));

  if (!flagdelay)
    socket_tcpnodelay(s); /* if it fails, bummer */

  if (pipe(pi) == -1) logmsg(WHO,111,FATAL,"unable to create pipe");
  if (pipe(po) == -1) logmsg(WHO,111,FATAL,"unable to create pipe");
  if (pi[0] == 7) {
    if (pipe(pt) == -1) logmsg(WHO,111,FATAL,"unable to create pipe");
    close(pi[0]); close(pi[1]);
    pi[0] = pt[0]; pi[1] = pt[1];
  }
  if (po[1] == 6) {
    if (pipe(pt) == -1) logmsg(WHO,111,FATAL,"unable to create pipe");
    close(po[0]); close(po[1]);
    po[0] = pt[0]; po[1] = pt[1];
  }

  switch (opt = fork()) {
    case -1:
      logmsg(WHO,111,FATAL,"unable to fork");
    case 0:
      break;
    default:
      close(pi[0]); close(po[1]);
      if (ssl_io(ssl,pi[1],po[0],progtimeout)) {
        logmsg(WHO,110,DROP,"unable to speak TLS");
        ssl_close(ssl);
        wait_pid(&wstat,opt);
        _exit(111);
      }
      ssl_close(ssl);
      if (wait_pid(&wstat,opt) > 0)
        _exit(wait_exitcode(wstat));
      _exit(0);
  }
  ssl_close(ssl); close(pi[1]); close(po[0]);

  if (flagsslenv && !ssl_client_env(ssl,0)) nomem();

  if (fd_move(6,pi[0]) == -1)
    logmsg(WHO,111,FATAL,"unable to set up descriptor 6");
  if (fd_move(7,po[1]) == -1)
    logmsg(WHO,111,FATAL,"unable to set up descriptor 7");
  sig_uncatch(sig_pipe);

  pathexec(argv);
  logmsg(WHO,111,FATAL,B("unable to run: ",*argv));

  return 0; /* never happens, but avoids compile warning */
}
